### PR TITLE
feat: use component library's dropdown component in learn 

### PR DESCRIPTION
**Pull request by ahmaxed**
This pr includes
- feat: make ui-components share same presets as learn
- feat: improve dropdown styles
- feat: add hot reloading for component library CSS
- feat: add component library to learn

Remaning tasks:
1 - extract types declarations from the component library and import them to learn.
2 - style the toggle button when toggled to look like the hover state using headless ui's internal state.
3 - include the linking scripts in package.json.
4 - build and publish the component library upon approval.

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
Hey @ahmadabdolsaheb what's the plan re: publishing? Are we going to publish to npm first or rework this to use the local code?

File: client/src/components/solution-display-widget/index.tsx [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
-import { Button, Dropdown, MenuItem } from '@freecodecamp/react-bootstrap';
+import { Button } from '@freecodecamp/react-bootstrap';
+import { Dropdown, MenuItem } from '@freecodecamp/ui';
 import React from 'react';
 import { useTranslation } from 'react-i18next';
 import { CompletedChallenge } from '../../redux/prop-types';
 import { getSolutionDisplayType } from '../../utils/solution-display-type';
 import './solution-display-widget.css';
+import '@freecodecamp/ui/dist/base.css';
```
Comment (Sboonny) [inline client/src/components/solution-display-widget/index.tsx:10]: works but need to a way to build auto build base.css, if we are going to merge as is

**Comment by ahmadabdolsaheb** [comment]
Once the tests pass we could do either of the followings:
1-Split this into two prs (learn and ui-components). Once the ui-components pr is merged, I could publish and then we could merge the learn one.
2-publish the component library from the pr and then merge this pr. 
Either way I am happy.

**Comment by ojeytonwilliams** [comment]
Both seem reasonable to me. Option 2 sounds like it might waste less time, so do you want to go with that?

**Comment by ahmadabdolsaheb** [comment]
Package published: https://www.npmjs.com/package/@freecodecamp/ui

**Comment by ahmadabdolsaheb** [comment]
Another round of reviews then merge?

**Comment by ojeytonwilliams** [comment]
@ahmadabdolsaheb it looks like there are a couple of issues. 

Firstly pnpm defaults to using the `@freecodecamp/ui` workspace. This is probably fine, but means we need to build it before linting etc.

Secondly, the types aren't quite working correctly. To fix this we'll need a `"types"` property in the package.json of ui-components, and point it at the types. Also we're not generating the declaration files any more. I could have sworn we did at one point!

**Comment by ahmadabdolsaheb** [comment]
The timeline jest tests need a little work. After that, it should be good to go.

**Comment by ahmadabdolsaheb** [comment]
Two tests need to be migrated to e2e:
1. certification settings.
2. timeline.

Either a multifile project needs to be added to the certified seed object or cypress needs to submit one so it shows up as the last one on the timeline.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @ahmadabdolsaheb the new dropdown looks great, but doesn't seem as friendly from an a11y perspective.

Right now, in production, it's possible to tab onto a dropdown, press enter to expand and then use the arrow keys to navigate between the options.

In this PR all of that is works except for the arrow key navigation, so it's not possible to make a selection.

I'd rather we delayed a bit than shipped a less accessible version.

File: client/tailwind.config.js
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply client/tailwind.config.js:1]: done.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams should be fixed.

**Review by ojeytonwilliams — APPROVED** [review]
It's looking great. Thanks for those final fixes, @ahmadabdolsaheb

**Comment by bbsmooth** [comment]
I'd like to take a look at this but I'm having trouble getting it to run with `pnpm run develop`. Getting the errors "Can't resolve '@freecodecamp/ui'" and "Can't resolve '@freecodecamp/ui/dist/base.css'". Is there a trick to this?

**Comment by ojeytonwilliams** [comment]
Hey @bbsmooth, I'm glad you spotted this - I should have pinged you!

> Getting the errors "Can't resolve '@freecodecamp/ui'" and "Can't resolve '@freecodecamp/ui/dist/base.css'".

At a guess, your pnpm version is too low. They changed the lockfile format recently and that seems to have broken things. Updating pnpm and `pnpm install`ing again should fix it.

Failing that, try removing node_modules and installing again. That shouldn't be necessary, though.

Failing _that_ @ me again and I'll see if I can figure out what's wrong.

**Comment by ahmadabdolsaheb** [comment]
Hmm, have you tried clean-and-develop?

**Comment by bbsmooth** [comment]
Thank you @ojeytonwilliams, the pnpm upgrade did the trick. I'll try to remember that next time. 

I'll put this through its paces and write up any notes soon.

**Comment by bbsmooth** [comment]
Here are my notes for the Get Help menu in the JS ADS challenges.

Tested with: http://localhost:8000/learn/javascript-algorithms-and-data-structures/basic-javascript/compound-assignment-with-augmented-subtraction50465

**Should be fixed:**

- A nice improvement is that now when the Get Help button is activated, focus is placed on the first item in the menu. Unfortunately, this breaks when using either the JAWS or NVDA screen reader. I will figure out why this is happening and let you know.

- Get Help -> Get a Hint: We should add sr-only text "opens in new window" to the end of the menu text here (this will be an improvement over current version on production). Should also consider adding the opens in new window icon that we use in the main menu.

- Get Help -> Watch a Video: When closing the video either by Esc key or 'x' close button, focus should be placed back on Get Help button (this is current behavior on production)

- Get Help -> Ask for Help: When closing the help dialog either by Esc key, 'x' close button, or Cancel button, focus should be placed back on Get Help button (this is current behavior on production)

**Nice to have but not required:**

- Allow the up/down arrow keys to wrap around in the menu (this is current behavior on production)

- In addition to the background color changing on focus of menu item, add the blue outline (using :focus-visible) as well for consistency (this is current behavior on production)

**Comment by ojeytonwilliams** [comment]
Thanks for the meticulous review, @bbsmooth. One thing I'm curious about is this: can we automate this kind of QA? At least now that you've pointed out what needs to be happening.

Some unit tests with react-testing-library's tools would be good, assuming they can catch the issues you described.

**Comment by bbsmooth** [comment]
More information on the issue with keyboard focus not automatically being placed on the first menu item when using JAWS/NVDA. This menu implementation is using `aria-activedescendant` to handle the focus management. Using the keyboard without a screen reader works properly. When the Get Help button is activated, the menu div is inserted into the DOM and the `aria-activedescendant` attribute (on the menu div) is set to the `id` of the first item in the menu. This is also (surprisingly) working properly when using the Orca screen reader in Linux.

But when using JAWS/NVDA, the `aria-activedescendant` attribute is NOT being set on the menu div when activating the Get Help button with the keyboard. I'm noticing that if you click on the Get Help button with the mouse that the `aria-activedescendant` attribute is also not set on the menu div either. So I'm assuming the Get Help button is attempting to determine whether the user used a keyboard to click the button or a pointer device, and only wants to set `aria-activedescendant` when using a keyboard. Unfortunately, whatever method the button is using to make this distinction isn't working when using JAWS/NVDA.

 I'm also finding this bug on the [headless ui menu example](https://headlessui.com/react/menu). So this seems like a bug with the component itself and not something we are doing. I guess I'll need to put together a codepen demo and open an issue in their github.

**Comment by ahmadabdolsaheb** [comment]
Thanks for the meticulous review, @bbsmooth. Just have a few questions regarding the dropdown menu implementation:

> Allow the up/down arrow keys to wrap around in the menu (this is current behavior on production)

Does this mean to have the icon wrap to the next line on smaller screens?

> Get Help -> Get a Hint: We should add sr-only text "opens in new window" to the end of the menu text here (this will be an improvement over current version on production). Should also consider adding the opens in new window icon that we use in the main menu.

I have no idea how to start implementing this, so pointers are appreciated.

File: client/src/templates/Challenges/components/tool-panel.tsx [hunk 2]
```diff
@@ -104,40 +101,26 @@ function ToolPanel({
...
-          >
-            {t('buttons.watch-video')}
+      <Dropdown dropup>
+        <Dropdown.Toggle dropup id={'get-help-dropdown'}>
+          {isMobile ? t('buttons.help') : t('buttons.get-help')}
+        </Dropdown.Toggle>
+        <Dropdown.Menu dropup>
+          {guideUrl ? (
+            <MenuItem href={guideUrl} target='_blank'>
+              {t('buttons.get-hint')}
```
Comment (bbsmooth) [inline client/src/templates/Challenges/components/tool-panel.tsx:111]:
```suggestion
              {t('buttons.get-hint')}{' '}
              <FontAwesomeIcon icon={faExternalLinkAlt} />
              <span className='sr-only'>, {t('aria.opens-new-window')}</span>
```

We should definitely add the sr-only text. The icon would be nice but isn't as essential if you don't like it.

File: client/src/templates/Challenges/components/tool-panel.tsx [hunk 3]
```diff
@@ -1,8 +1,5 @@
-import {
-  Button,
-  DropdownButton,
-  MenuItem
-} from '@freecodecamp/react-bootstrap';
+import { Button } from '@freecodecamp/react-bootstrap';
+import { Dropdown, MenuItem } from '@freecodecamp/ui';
```
Comment (bbsmooth) [inline client/src/templates/Challenges/components/tool-panel.tsx:2]:
```suggestion
import { Dropdown, MenuItem } from '@freecodecamp/ui';
import { faExternalLinkAlt } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
```
We'll need these if we add the icon to the Get a Hint menu item.

**Comment by bbsmooth** [comment]
> > Allow the up/down arrow keys to wrap around in the menu (this is current behavior on production)
> 
> Does this mean to have the icon wrap to the next line on smaller screens?

Sorry, I might not have been as clear as I could have been. If focus is currently on the top item in the menu, pressing the up arrow key moves focus to the bottom item. Likewise, if focus is currently on the bottom item, pressing the down arrow key moves focus to the top item. 

This is not required keyboard functionality, just nice to have, and the menu on production currently has it.

**Comment by ahmadabdolsaheb** [comment]
It seems like bootstrap modal finds the last active element before the modal is open and set .focus() to that element when the modal is closed. 
Resources: [modal](https://github.com/react-bootstrap/react-bootstrap/blob/bs3-dev/src/Modal.js), [baseModal](https://github.com/react-bootstrap/react-overlays/blob/0.8-stable/src/Modal.js#L3), [activeElements](https://github.com/react-bootstrap/dom-helpers/blob/master/src/activeElement.ts)

**Comment by bbsmooth** [comment]
> I was wondering if we could move forward without the followings since they need to be resolved on headlessui's side?

Yes, definitely. I don't think either of these issues are blockers. The arrow wrap is definitely not (just nice to have). The lack of focus on the first menu item when using NVDA/JAWS is slightly more annoying but won't keep these users from using the menu. I plan on creating an issue for this when I have a few minutes to put together a good demo.

**Comment by bbsmooth** [comment]
Well, I found another "issue" but I don't think there is anything we can do about it (this seems to be a bug in headlessui). When you open the menu with a keyboard and it places focus on the first menu item, if you then hit the Tab key it closes the menu and places the focus on the next focusable item on the page after the "Get Help" button (which is the JS editor in this case). That is the expected behavior. But if you do the same thing but use `Shift + Tab` then it still places you in the JS editor instead of going backwards and placing the focus on the "Reset this lesson" button, which is the previous focusable element before the "Get Help" button. I guess this is another issue to open with headlessui. Shouldn't stop us from going forward.

And just one very minor nitpick under our control. If you don't think it is worthy of fixing now, that's fine. If it really bugs me I can open my own PR.

focus-outline-cut-off

As you can see in the above pic, when zoomed in, the left side of the focus outline is cut off so only 1px is showing. I think setting `outline-offset: -3px` on these menu items for focus-visible/hover will solve that problem. We have this same issue for the Help/Reset/Run buttons at the bottom, but that's perhaps a little out of scope for this PR so I will open a new PR for those (unless you want to deal with them now).

Otherwise, the other fixes look good and this is ready to go.

File: tools/ui-components/src/normalize.css [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  height: 0;
+}
+pre {
+  overflow: auto;
+}
+code,
+kbd,
+pre,
+samp {
+  font-family: monospace, monospace;
```
[the remark this answers is not in the document]
Comment (ahmadabdolsaheb) [reply tools/ui-components/src/normalize.css:99]: No it should not.

**Review by bbsmooth — COMMENTED** [review]
LGTM.

**Review by moT01 — APPROVED** [review]

**Pull request merged**